client/extensions/hello-dolly/state/reducer.js

```diff
@@ -0,0 +1,25 @@
+/**
+ * Internal dependencies
+ */
+import { HELLO_DOLLY_NEXT_LYRIC } from './action-types';
+import {
+	ROUTE_SET,
+	SECTION_SET,
+	SITE_SETTINGS_SAVE,
+} from 'state/action-types.js';
+
+export default function lyricIndex( state = 0, action ) {
+	switch ( action.type ) {
+		case ROUTE_SET:
+		case SECTION_SET:
+		case SITE_SETTINGS_SAVE:
+		case HELLO_DOLLY_NEXT_LYRIC:
+			return advanceToNextLyric( state );
+		default:
+			return state;
+	}
+}
+
+function advanceToNextLyric( index ) {
+	return index + 1;
```

Comment:
Pro tip: `hello-dolly` could work without similarly to how it works right now but without its own reducer if it consumed `state.ui.actionLog` — see Guided Tours' [`ARCHITECTURE.md`](https://github.com/Automattic/wp-calypso/blob/master/client/layout/guided-tours/docs/ARCHITECTURE.md).